chalk-integration/src/lowering.rs

```diff
@@ -354,6 +372,38 @@ impl LowerProgram for Program {
                 Item::Clause(ref clause) => {
                     custom_clauses.extend(clause.lower_clause(&empty_env)?);
                 }
+                Item::OpaqueTyDefn(ref opaque_ty) => {
+                    if let Some(&value) = impl_trait_ids.get(&opaque_ty.identifier.str) {
+                        let parameter_kinds = opaque_ty
+                            .parameter_kinds
+                            .iter()
+                            .map(|k| k.lower())
+                            .collect::<Vec<_>>();
+
+                        let binders = empty_env
+                            .in_binders(parameter_kinds, |env| opaque_ty.ty.lower(&env))?;
+
+                        impl_trait_data.insert(
+                            value,
+                            Arc::new(ImplTraitDatum {
+                                impl_trait_id: value,
+                                bounds: opaque_ty
+                                    .bounds
+                                    .lower(&empty_env)?
+                                    .iter()
+                                    .flat_map(|qil| {
+                                        qil.into_where_clauses(
+                                            empty_env.interner(),
+                                            chalk_ir::TyData::BoundVar(0)
```

Comment:
Something seems suspicious here:

(a) I expect these where clauses to be lowering inside of the `parameter_kinds` binder, first off. This is because you may have something like

```rust
type Foo<T> = impl Iterator<Item = T>
```

(b) Using `BoundVar` here seems a bit surprising because we're kind of missing a *binder*. If we were taking the approach I suggested above, though, of indeed adding a `BoundVar` to represent "the hidden type" (or "the opaque type placeholder" as case may be), that would be an additional binder we push on. Hmm, we should probably talk this out a bit more on Zulip.